src/term_arena.rs

```diff
@@ -368,6 +377,8 @@ pub enum Term {
     Int(i64),
     /// Prune all further alternatives down to the level of the goal that the cut appeared in.
     Cut,
+    /// A resolver-defined constraint (WIP: less-than)
+    Constraint(i64),
```

Comment:
In the final version, anything resolver defined should just be represented as an `AppTerm` I'd say. Only reason I turned `Cut` into its own term type is because it has special implementation support in the solver core.

The surface language doesn't necessarily need to stick to alphanumeric names and prefix-notation for those special app-terms either. It would be fine to e.g. parse `X #< 4` in the surface language as an AppTerm of the form `#< (X, 4)`

Reply:
I did this mostly because it's more comfortable to have a Rust structure to operate on, rather than parse the terms AST every time a constraint needs to be dealt with.

If there was a generic way to add a constraint of an arbitrary type, I'd have used that.